cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,468 @@
+                pg_loss = torch.max(pg_loss1, pg_loss2).mean()
+
+                entropy_loss = entropy.mean()
+                policy_loss = pg_loss - args.ent_coef * entropy_loss
+
+                policy_optimizer.zero_grad()
+                policy_loss.backward()
+                nn.utils.clip_grad_norm_(agent_policy.parameters(), args.max_grad_norm)
+                policy_optimizer.step()
+
+            if args.target_kl is not None:
+                if approx_kl > args.target_kl:
+                    break
+
+        # Value network optimization
+        for epoch in range(args.value_update_epochs):
+            np.random.shuffle(b_inds)
+            for start in range(0, args.epoch_size, args.value_batch_size):
+                end = start + args.value_batch_size
+                mb_inds = b_inds[start:end]
+
+                # Value loss
+                _, newvalue = agent_value(b_obs[mb_inds])
+                v_loss = 0.5 * (newvalue - b_returns[mb_inds]).square().mean()
+
+                value_optimizer.zero_grad()
+                v_loss.backward()
+                nn.utils.clip_grad_norm_(agent_value.parameters(), args.max_grad_norm)
+                value_optimizer.step()
+
+        y_pred, y_true = b_values.cpu().numpy(), b_returns.cpu().numpy()
+        var_y = np.var(y_true)
+        explained_var = np.nan if var_y == 0 else 1 - np.var(y_true - y_pred) / var_y
+
+        # TRY NOT TO MODIFY: record rewards for plotting purposes
+        writer.add_scalar("charts/policy_learning_rate", policy_optimizer.param_groups[0]["lr"], global_step)
+        writer.add_scalar("charts/value_learning_rate", value_optimizer.param_groups[0]["lr"], global_step)
+        writer.add_scalar("charts/distill_learning_rate", distill_optimizer.param_groups[0]["lr"], global_step)
+        writer.add_scalar("losses/value_loss", v_loss.item(), global_step)
+        writer.add_scalar("losses/policy_loss", pg_loss.item(), global_step)
+        writer.add_scalar("losses/distill_loss", distill_loss.item(), global_step)
+        writer.add_scalar("losses/entropy", entropy_loss.item(), global_step)
+        writer.add_scalar("losses/old_approx_kl", old_approx_kl.item(), global_step)
+        writer.add_scalar("losses/approx_kl", approx_kl.item(), global_step)
+        writer.add_scalar("losses/clipfrac", np.mean(clipfracs), global_step)
+        writer.add_scalar("losses/explained_variance", explained_var, global_step)
+        print("SPS:", int(global_step / (time.time() - start_time)))
+        writer.add_scalar("charts/SPS", int(global_step / (time.time() - start_time)), global_step)
+        writer.add_scalar("env_step", env_step, global_step)
+
+    envs.close()
+    writer.close()
+
+    if args.track:
+        # save final model
+        torch.save(agent_policy.state_dict(), "agent_policy.pt")
+        torch.save(agent_value.state_dict(), "agent_value.pt")
+        wandb.save("agent_policy.pt")
+        wandb.save("agent_value.pt")
```

Comment:
Out of curiosity, what is the use case of this?

Reply:
removed

Reply:
How would you generate gameplay videos, though? Through the gym wrappers? This might be a legitimate use case for model saving.